File: runtime/elixir/test/runtime_http_client_test.exs
defmodule RuntimeHttpClientTest do
  use ExUnit.Case, async: true

  alias RuntimeHttpClient
  alias RuntimeTypes.HttpRequest

  test "build_url/2 without query" do
    config = %{base_url: "https://api.example"}

    req = %HttpRequest{
      method: "GET",
      path: "/items",
      query: %{},
      headers: [],
      body: ""
    }

    assert "https://api.example/items" == RuntimeHttpClient.build_url(config, req)
  end

  test "build_url/2 with query" do
    config = %{base_url: "https://api.example"}

    req = %HttpRequest{
      method: "GET",
      path: "/items",
      query: %{"verbose" => "true"},
      headers: [],
      body: ""
    }

    url = RuntimeHttpClient.build_url(config, req)
    assert String.contains?(url, "?verbose=true")
  end

  test "build_request/2" do
    config = %{base_url: "https://api.example"}

    req = %HttpRequest{
      method: "POST",
      path: "/items",
      query: %{},
      headers: [{"Content-Type", "application/json"}],
      body: ~s({"name":"item"})
    }

    client_req = RuntimeHttpClient.build_request(config, req)
    assert client_req.method == :post
    assert client_req.url == "https://api.example/items"
    assert client_req.headers == req.headers
    assert client_req.body == req.body
  end

  test "content_type/1 defaults" do
    assert "application/octet-stream" == RuntimeHttpClient.content_type([])
  end

  test "content_type/1 from header" do
    headers = [{"Content-Type", "application/json"}]
    assert "application/json" == RuntimeHttpClient.content_type(headers)
  end
end

File: runtime/erlang/test/runtime_http_client_test.erl
-module(runtime_http_client_test).

-include_lib("eunit/include/eunit.hrl").
-include("runtime_types.hrl").
-include("runtime_http_client.hrl").

build_url_without_query_test() ->
    Config = #{base_url => <<"https://api.example">>},
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/items">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    ?assertEqual(
        <<"https://api.example/items">>,
        runtime_http_client:build_url(Config, Req)
    ).

build_url_with_query_test() ->
    Config = #{base_url => <<"https://api.example">>},
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/items">>,
        query = #{<<"verbose">> => <<"true">>},
        headers = [],
        body = <<>>
    },
    Url = runtime_http_client:build_url(Config, Req),
    ?assertNotEqual(nomatch, binary:match(Url, <<"?verbose=true">>)).

build_request_test() ->
    Config = #{base_url => <<"https://api.example">>},
    Req = #http_request{
        method = <<"POST">>,
        path = <<"/items">>,
        query = #{},
        headers = [{<<"Content-Type">>, <<"application/json">>}],
        body = <<"{\"name\":\"item\"}">>
    },
    ClientReq = runtime_http_client:build_request(Config, Req),
    ?assertEqual(post, ClientReq#http_client_request.method),
    ?assertEqual(<<"https://api.example/items">>, ClientReq#http_client_request.url),
    ?assertEqual(Req#http_request.headers, ClientReq#http_client_request.headers),
    ?assertEqual(Req#http_request.body, ClientReq#http_client_request.body).

content_type_defaults_test() ->
    ?assertEqual(
        <<"application/octet-stream">>,
        runtime_http_client:content_type([])
    ).

content_type_from_header_test() ->
    Headers = [{<<"Content-Type">>, <<"application/json">>}],
    ?assertEqual(<<"application/json">>, runtime_http_client:content_type(Headers)).
